torchvision/transforms/functional.py

```diff
@@ -82,6 +82,54 @@ def to_tensor(pic):
         return img
 
 
+def as_tensor(pic):
+    """Convert a ``PIL Image`` or ``numpy.ndarray`` to tensor of same type.
+
+    See ``AsTensor`` for more details.
+
+    Args:
+        pic (PIL Image or numpy.ndarray): Image to be converted to tensor.
+
+    Returns:
+        Tensor: Converted image.
+    """
+    if not(_is_pil_image(pic) or _is_numpy(pic)):
+        raise TypeError('pic should be PIL Image or ndarray. Got {}'.format(type(pic)))
+
+    if _is_numpy(pic) and not _is_numpy_image(pic):
+        raise ValueError('pic should be 2/3 dimensional. Got {} dimensions.'.format(pic.ndim))
+
+    if isinstance(pic, np.ndarray):
+        # handle numpy array
+        if pic.ndim == 2:
+            pic = pic[:, :, None]
+
+        img = torch.from_numpy(pic.transpose((2, 0, 1)))
+        return img
+
+    if accimage is not None and isinstance(pic, accimage.Image):
+        nppic = np.zeros([pic.channels, pic.height, pic.width], dtype=np.float32)
+        pic.copyto(nppic)
+        return torch.from_numpy(nppic)
+
+    # handle PIL Image
+    if pic.mode == 'I':
+        img = torch.from_numpy(np.array(pic, np.int32, copy=False))
+    elif pic.mode == 'I;16':
+        img = torch.from_numpy(np.array(pic, np.int16, copy=False))
+    elif pic.mode == 'F':
+        img = torch.from_numpy(np.array(pic, np.float32, copy=False))
+    elif pic.mode == '1':
+        img = torch.from_numpy(np.array(pic, np.uint8, copy=False))
+    else:
+        img = torch.ByteTensor(torch.ByteStorage.from_buffer(pic.tobytes()))
```

Comment:
@pmeier Do you have an opinion about this line though?

I wasn't sure how safe it is to have the image always converted to a torch.ByteTensor as a final catch all.  It does handle the "L" mode and a few other missing modes from Pillow but I wanted a second opinion.

Reply:
I'm not sure why it is there anyway. Is there a difference to what we do in 123?

Reply:
```python
import numpy as np
import timeit
import torch

from PIL import Image

x = Image.fromarray(np.ones((256, 256, 3), dtype=np.uint8))

def fn1(x):
    return torch.from_numpy(np.array(x, np.uint8, copy=False))

def fn2(x):
    return torch.ByteTensor(torch.ByteStorage.from_buffer(x.tobytes()))

number = 10000

fn1_time = timeit.timeit(lambda: fn1(x), number=number)
print(f"fn1 time: {fn1_time /number * 1e6:.2f} µs")
#fn1 time: 85.38 µs

fn2_time = timeit.timeit(lambda: fn2(x), number=number)
print(f"fn2 time: {fn2_time /number * 1e6:.2f} µs")
#fn1 time: 85.38 µs

rel_diff = fn1_time / fn2_time - 1.0
print(f"rel. diff.: {rel_diff:+.1%}")                                                        
#rel. diff.: -9.2%
```

I can't really say when one would want to use the second case over the first especially given that it is slower.  I will defer to someone on pytorch team to take a look though to let me know if it can be safely removed.

Reply:
You could locally merge the two cases in `to_tensor` and run the tests. If nothing pops up, this is just an artifact from an organically grown function and most likely can be removed. If something pops up you have a place to start investigating why this case is special.

Reply:
This is due to historical artifacts.

Indeed, there was no `from_numpy` in the initial version of PyTorch, and the `torch.ByteTensor(torch.ByteStorage.from_buffer(x.tobytes()))` was the efficient way of doing it back then without introducing a dependency on numpy, but it's not needed (and not recommended) to be done anymore, as we how depend on numpy in pytorch as well (and in torchvision)

What I would recommend doing though is to use `torch.as_tensor` instead of `from_numpy`. It's more generic than `from_numpy`, and I'd say the recommended way going forward